Client scheduler pr branch

akashkumar2296 opened the pull request.

akashkumar2296 committed:
Client Scheduler PR ready

akashkumar2296 committed:
Run Algorithm function deleted

akashkumar2296 commented:
When clicking "Schedule Tours" button, the algorithm is run using user_profile.favorites.all()

This doesn't necessarily hold the scheduler to 1-6 homes but all the favorites of the user. We should change the functionality so that theres a way for the user to only favorite 1-6 before clicking "Schedule tours", or we should run the algorithm separately from the visitList.html page. But as far as I saw, theres no way to change the visit list so the algorithm is being run on all favorites

tiusty commented on cocoon/scheduler/clientScheduler/test_client_scheduler_algorithm.py:
```diff
@@ -0,0 +1,42 @@
+from client_scheduler import clientSchedulerAlgorithm
+
+
+class TestClientSchedulerAlgorithm(TestCase):
```
I am not convinced that the alogrithm works at all cases by seeing these tests.
Tests for if the number of nodes is less than 5 would be good, i.e 1, 2 3 etc. Also tests verifying that is working correctly with more configurations. I.e if 2 homes are closer together then 2 homes far away. Then maybe a group of 3 homes close and a group of 2 homes far etc etc.

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,46 @@ def survey_result_rent(request, survey_url=""):
+
+    # Create a string so that it can be passed into ContentFile, which is readable in the FileSystem operation
+    s = ""
+    for item in interpreted_route:
```
An imporant part of the route is the time it takes to go between each node. Can you append the edge weight to the itinerary file? Also we need to make sure to add time for each home. So add 20 minutes to ethe estimated to for each home that is stopped at

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,46 @@ def survey_result_rent(request, survey_url=""):
+
+    # Update Itinerary Model
+    # File name is unique based on user and current time (making it impossible to have duplicates)
+    itinerary_model = ItineraryModel(client=request.user, selected_start_time=datetime.now())
```
I would make sure that an itinerary model doesn't already exist before creating one. Otherwise a new itinerary model is created on each page load

tiusty commented on cocoon/scheduler/clientScheduler/client_scheduler.py:
```diff
@@ -0,0 +1,67 @@
+class ClientScheduler(object):
+
+    def __init__(self):
+        self.algorithm = clientSchedulerAlgorithm()
```
can you have the ClientScheduler inherit the cilentScheduler Algorithm instead of a member data?

tiusty commented on cocoon/scheduler/clientScheduler/base_algorithm.py:
```diff
@@ -0,0 +1,45 @@
+        returning its index
+        :param homes_matrix: The matrix calcualted using DistanceWrapper() with distances between every pair of homes in
+        favorited list
+        :return: starting_point: the Global Minimum. This is the smallest possible distance, so the algorithm will start
```
can add you add the data type for the arguments and the return type like I did in my function. I.e (int). Do it for the other functions as well

Commit:
New Branch with updated Develop

Commit:
Fixed Client Scheduler Algorithm inheritance

Commit:
Init + testing directory

Commit:
Init + testing directory

Commit:
Edge weights done (with 20 minute lag)
Create Itinerary model only if it doesn't exist
ClientScheduler inherits ClientSchedulerAlgorithm
Datatypes added to return types in comments
Tests folder added to clientScheduler

Commit:
fixed merge conflicts

Commit:
Missed a edge weight in unit testing

tiusty commented on .gitignore:
```diff
@@ -52,3 +52,7 @@ db.sqlite3
 # The log file generated #
 logCocoon.log
+
+#DS store (mac)
```
What is this for?

akashkumar2296 replied:
Its part of the mac OS, .DS_Store should always be gitignored

tiusty commented on cocoon/scheduler/clientScheduler/client_scheduler.py:
```diff
@@ -0,0 +1,73 @@
+
+    def __init__(self):
+        self.wrapper = DistanceWrapper()
+        clientSchedulerAlgorithm.__init__(self)
```
this isn't the correct way to call a parents constructor function,
should be:
`super().__init__()`

tiusty commented on cocoon/scheduler/clientScheduler/client_scheduler.py:
```diff
@@ -0,0 +1,73 @@
+
+            home_one_distances = []
+
+            result_distance_wrapper = self.wrapper.get_durations_and_distances(origins=[home_one],
```
you don't want to call the get_durations_and_distances function, use the new retrieve_exact_commutes instead and pass in driving manually to make it clear what it is doing

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
+        itinerary_model = ItineraryModel.objects.get(client=request.user)
+
+    except ItineraryModel.DoesNotExist:
+        itinerary_model = ItineraryModel(client=request.user, selected_start_time=datetime.now())
```
we do not want the selected_start_time to be set. The selected start time is set when the agent selects a start time based on all the available start times

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
+    except ItineraryModel.DoesNotExist:
+        itinerary_model = ItineraryModel(client=request.user, selected_start_time=datetime.now())
+
+    itinerary_model.itinerary.save(os.path.basename("itinerary_route_" + str(request.user)) + "_" + str(datetime.now()),
```
The file path should not be generated here, it should be generated in the model's file like I did for the MLSPIN requester images. I.e create a helper a callable that generates the file path

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
+
+    itinerary_model.itinerary.save(os.path.basename("itinerary_route_" + str(request.user)) + "_" + str(datetime.now()),
+                                   ContentFile(s))
+    itinerary_model.save()
```
I don't see the homes getting added to the itinerary. The homes that are used to create the route then needs to be saved back into the itinerary so we have a record of what the homes were

tiusty commented on cocoon/scheduler/clientScheduler/base_algorithm.py:
```diff
@@ -0,0 +1,57 @@
+        :return: (int): starting_point - the Global Minimum. This is the smallest possible distance, so the algorithm will start
+        from this address
+        '''
+        min_time = 1000000000
```
a more robust solution would be to set the min_time to house_matrix[0][1], the first non-zero distance

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
 @login_required
 def visit_list(request):
+    """
+        AJAX request used to run the client scheduler algorithm and also render the visitList.html page. This is
```
This comment is not true, this isn't an AJAX request. Anyways when you pull in develop this function will get rewritten anyway.

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
+        the main driver function that uses all methods of the client scheduler algorithm and outputs the itinerary
+        list file to s3 and populates an instance of the itinerary model.
+        :param request: The HTTP request
+        :return: HttpResponse if everything goes well. It returns a lot of context variables.
```
Lol

tiusty commented on cocoon/survey/views.py:
```diff
@@ -254,17 +264,56 @@ def survey_result_rent(request, survey_url=""):
 
-    # Since the page is loading, update all the signed documents to see if the status has changed
+    # Run the client scheduler algorithm
+    homes_list = []
```
Views are not meant to have a lot of functionality. I updated the renting survey/survey result to reflect that. If you could put this into a def run() function in the client scheduler that would be great. Then you only pass in a few arguments and it does it within the client_scheduler module instead of in the view.

Commit:
PR changes

tiusty committed:
Merge remote-tracking branch 'origin/develop' into client_scheduler_pr_branch

# Conflicts:
#	cocoon/survey/views.py

tiusty committed:
Made fix to context variable

Commit:
Merge conflict local

Commit:
Merge branch 'client_scheduler_pr_branch' of https://github.com/tiusty/Cocoon into client_scheduler_pr_branch

Commit:
TODO: (PR Comment) setting homes for itinerary model
(Unit Testing) Mocking the retrieve_exact_commute function

tiusty commented on cocoon/scheduler/clientScheduler/client_scheduler.py:
```diff
@@ -0,0 +1,114 @@
+
+class ClientScheduler(clientSchedulerAlgorithm):
+
+    def __init__(self):
```
You actually can just delete the __init__() function since it can just inherit it from the base class. It isn't doing anything anymore.

tiusty commented on cocoon/scheduler/models.py:
```diff
@@ -135,7 +138,9 @@ def unschedule_itinerary(self, **kwargs):
         email.send()
         
-        
+def itinerary_directory_path(user):
```
Actually I meant that it should be a callable for the file_field I can show you at the meeting

Commit:
(TODO): unit testing

Commit:
Testing Done - PR comments complete

Commit:
Testing Done - PR comments complete

tiusty commented on cocoon/scheduler/clientScheduler/tests/test_client_scheduler_wrapper.py:
```diff
@@ -0,0 +1,156 @@
+        :return:
+        '''
+
+        homes_list = ["home1", "home2", "home3"]
```
Don't you actually want to test with real homes since that is what the function is actually going to be called with? Also add the assertcalledwith functions to assert that the retrieve exact commutes are called properly with the right arguments

Commit:
gitingore - mac

Commit:
retrieve_exact_commute temp change

Commit:
retrieve_exact_commute_commute temp change

Commit:
retrieve_exact_commute temp change
